Create an app configuration

MarcelRobitaille opened the pull request:
Fixes #1067

Move the AppNavigationSettings from the sidebar to a AppSettingsDialog modal that covers the screen.

This is what it should look like:

![image](https://user-images.githubusercontent.com/8503756/196308672-f745e419-eecf-46dc-b97e-cd7e7b81cba2.png)

This makes it easy for us to add more configuration options for things like notifications and alarms without cluttering the sidebar.

MarcelRobitaille commented:
@christianlupus That's weird, running `npm run styleling` locally doesn't warn about anything for me. I ran prettier, eslint, and styleling before pushing. BTW, I'm getting a deprecation warning with eslint now.

christianlupus commented:
Are you on the latest styleling config? There was a PR recently #1254.

christianlupus commented on src/components/SettingsDialog.vue:
```diff
@@ -1,19 +1,24 @@
-                            :title="t('cookbook', 'Rescan library')"
-                            @click="$emit('reindex')"
-                        />
+                        <NcButton @click="$emit('reindex')">
```
```suggestion
                        <Button @click="$emit('reindex')">
```

christianlupus commented on src/components/SettingsDialog.vue:
```diff
@@ -1,19 +1,24 @@
+                                <ReloadIcon v-else />
+                            </template>
+                            {{ t("cookbook", "Rescan library") }}
+                        </NcButton>
```
```suggestion
                        </Button>
```

christianlupus commented on src/components/SettingsDialog.vue:
```diff
@@ -64,35 +78,41 @@
-            type: Boolean,
-            default: false,
-        },
+        NcButton,
```
```suggestion
        Button,
```

christianlupus commented on src/components/SettingsDialog.vue:
```diff
@@ -64,35 +78,41 @@
+
+import AppSettingsDialog from "@nextcloud/vue/dist/Components/AppSettingsDialog"
+import AppSettingsSection from "@nextcloud/vue/dist/Components/AppSettingsSection"
+import NcButton from "@nextcloud/vue/dist/Components/Button"
```
```suggestion
import Button from "@nextcloud/vue/dist/Components/Button"
```

christianlupus commented on src/components/SettingsDialog.vue:
```diff
@@ -1,19 +1,24 @@
-                            :title="t('cookbook', 'Rescan library')"
-                            @click="$emit('reindex')"
-                        />
+                        <NcButton @click="$emit('reindex')">
```
The button is causing a JS exception on my machine when clicking it. Can you confirm?

MarcelRobitaille replied:
I think this should be fixed in c9d37206. I had to replace `$emit('reindex')` with `reindex` because now that method is in the same component.

christianlupus replied:
Does no longer happen on my machine. So it seems to be fixed in the meantime.

christianlupus reviewed (changes requested):
The PR has some problems yet. The component `NcButton` does not yet exist. It is part of the `@nextcloud/vue` v6 package. We are stuck with v5. Like this, the button is non-functional.

When changing the button, I get an error in the JS console:
![grafik](https://user-images.githubusercontent.com/8202922/196407176-3b9e1cea-3b94-4f10-a5b2-6433065c1daf.png)

Additionally, I cannot change the folder of the cookbook. I cannot navigate in the modal dialog to select a folder.

Can you confirm these issues?

MarcelRobitaille commented:
@christianlupus I called it `NcButton` and not `Button` because I got an error something like `Button is a reserved word in HTML`. There shouldn't be a problem with `import NcButton from '@nextcloud/vue/dist/Components/Button`, is there? It's fewer things to change when we finally migrate too.

I will have a look at resolving the other problems

MarcelRobitaille commented:
Here is the error when renaming `NcButton` to `Button`:

```
/home/marcel/code/cookbook/src/components/SettingsDialog.vue
  101:9  error  Name "Button" is reserved in HTML  vue/no-reserved-component-names
```

I think that is the cause of the errors in the screenshot you posted.

christianlupus commented:
> @christianlupus I called it `NcButton` and not `Button` because I got an error something like `Button is a reserved word in HTML`. There shouldn't be a problem with `import NcButton from '@nextcloud/vue/dist/Components/Button`, is there? It's fewer things to change when we finally migrate too.

Ahh, I am sorry, I missed that line. I assumed you were using `import {NcButton} from ...`. You are perfectly fine with that. Sorry for bothering you.

> Thanks for reviewing smile.
You are welcome.

Please tell me when you finished with the entire PR. Thanks!

MarcelRobitaille commented:
No worries.

I fixed the problem with the reindex button.

I am not sure what is going on with the file picker. I added this code in the developer console, and once the filepicker dialog is open, no events are fired at all:

```js
document.addEventListener('click', e => console.log(e))
```

I'm not really sure where this global `OC` is coming from. I can't find any documentation for it. I don't really like this kind of magic API, I prefer to import what I'm using explicitly. It's different from [this](https://github.com/nextcloud/nextcloud-dialogs/blob/master/lib/filepicker.ts), right?

I was not able to reproduce the error in your screenshot. Did you use the code exactly as it is here or did you replace `NcButton` with `Button` locally?

christianlupus commented:
> I'm not really sure where this global `OC` is coming from. I can't find any documentation for it. I don't really like this kind of magic API, I prefer to import what I'm using explicitly. It's different from [this](https://github.com/nextcloud/nextcloud-dialogs/blob/master/lib/filepicker.ts), right?

I do not know for sure. It comes from [here](https://github.com/nextcloud/cookbook/blob/296891ebf34f8b35fc691cd9cd0882f1b89a4afd/src/main.js#L38). I have read of it in the past but I do no longer find it.

> I was not able to reproduce the error in your screenshot. Did you use the code exactly as it is here or did you replace `NcButton` with `Button` locally?

Yes, I changed that locally. I have to retest it.

MarcelRobitaille commented:
I found the function: https://github.com/nextcloud/server/blob/master/core/src/OC/dialogs.js#L229-L572

It seems like a bug, but I'm not sure whether to report it to @nextcloud/vue or nextcloud/server.

MarcelRobitaille commented:
After hours of debugging, I found that this bug is not present in @nextcloud/vue <= 5.3.1. We could downgrade to this version until we move to 7.0.0, or we could try to upstream a patch and have it backported to 5.4.1.

MarcelRobitaille commented:
My guess is that the cause is https://github.com/nextcloud/nextcloud-vue/pull/2654. It seems like it was already [fixed in later versions](https://github.com/nextcloud/nextcloud-vue/pull/3020). Should we request a backport?

christianlupus commented:
I think a backport might be the cleanest solution.

Let's see, if @nextcloud/vuejs do know something about this issue. Can anyone of you confirm that there might be a problem and there is in fact a chance this can be fixed with the 5.x branch?

christianlupus commented:
ATM there are some commits that need fixing. I will wait and set this PR to be a draft. You can mark as ready for review once, you are done, @MarcelRobitaille.

MarcelRobitaille commented:
I rebased the NC25 changes. Now we're getting hung up on this open issue: https://github.com/nextcloud/nextcloud-vue/issues/3386

christianlupus commented:
:+1: for the rebase, :-1: for the upstream bug.

MarcelRobitaille commented:
The upstream issue was closed, but the fixes still haven't made it into a stable version.

MarcelRobitaille commented:
nextcloud/vue 7.10.0 was recently released, which fixes the "too much recursion" issue. I've just tested it and now everything seems to be working. The settings modal works and I can interact with the "Recipe folder" modal.

christianlupus reviewed (approved):
Seems good to me now. LGTM
